File: 2024/day01.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <iterator>
#include <cstdint>
#include <cstring>
#include <cmath>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <algorithm>
#include <numeric>

#include "utils.h"

using namespace std;

const string test(
"3   4\n"
"4   3\n"
"2   5\n"
"1   3\n"
"3   9\n"
"3   3"
);

bool DEBUG = false;
bool TEST = false;

int part1 (vector<int> & one, vector<int> & two )
{
    int sumx = 0;
    for( int i = 0; i < one.size(); i++ )
        sumx += abs(one[i]-two[i]);
    return sumx;
}

int part2 (vector<int> & one, vector<int> & two )
{
    int sumx = 0;
    for( int i = 0; i < one.size(); i++ )
        sumx += one[i] * count(two.begin(), two.end(), one[i] );
    return sumx;
}


int main( int argc, char ** argv )
{
    string name = *argv;
    while( *++argv )
    {
        string arg(*argv);
        if( arg == "debug")
            DEBUG = true;
        else if( arg =="test")
            TEST = true;
    }


    stringstream input;
    if( TEST )
        input << test;
    else
        input << ifstream("day01.txt").rdbuf();

    vector<int> data{
        istream_iterator<int>(input),
        istream_iterator<int>()
    };

    vector<int> one;
    vector<int> two;
    for( int i = 0; i < data.size(); i+=2 )
    {
        one.push_back( data[i] );
        two.push_back( data[i+1] );
    }
    sort( one.begin(), one.end() );
    sort( two.begin(), two.end() );

    cout << "Part 1: " << part1(one,two) << "\n";
    cout << "Part 2: " << part2(one,two) << "\n";
}

File: 2024/day01.go
package main

import (
	"fmt"
	"sort"

	"aoc/tools"
	_ "embed"
)

var DEBUG bool = false
var TEST bool = false
var test = `
3   4
4   3
2   5
1   3
3   9
3   3`[1:]

//go:embed day01.txt
var live string

func part1(one []int, two []int) int {
	sumx := 0
	for i, p1 := range one {
		sumx += tools.AbsDiffInt(p1, two[i])
	}
	return sumx
}

func part2(one []int, two []int) int {
	sumx := 0
	for _, p1 := range one {
		sumx += p1 * tools.Count(two, p1)
	}
	return sumx
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)
	data := tools.Parse[int](input)

	var one []int
	var two []int
	for _, row := range data {
		one = append(one, row[0])
		two = append(two, row[1])
	}

	sort.Ints(one)
	sort.Ints(two)
	fmt.Println("Part 1:", part1(one, two))
	fmt.Println("Part 2:", part2(one, two))
}
